### PR TITLE
Chore patches

### PR DESCRIPTION
**Pull request by jannisborn**
Minor fixes:

- #101 Pytorch import order fixed via `isort` preferences
- Added a new default SMILESLanguage dictionary (that one has seen GDSC, CCLE, BindingDB, ChEMBL (2M), VDJ and also ZINC (1.8 Billion SMILES). Should be sufficient to capture almost all tokens that can occur based on the RegExp)
- Refactored the package metadata naming to be more generic
- Some improved error handling and other housekeepping

### CODE REVIEW

**Comment by jannisborn** [comment]
Fixing also 2 more issues:
#99 
#84

File: .travis/Dockerfile [hunk 1]
```diff
@@ -6,4 +6,6 @@ RUN pip3 install --no-cache-dir -r requirements.txt
 # copy pytoda
 COPY . .
 RUN pip3 install --no-deps .
+RUN apt-get update
+RUN apt-get install python-pip curl git -y
```
Comment (C-nit) [inline .travis/Dockerfile]: these should happen at the top of a Dockerfile, i.e. after the FROM
I'm curious, why is it needed?

File: pytoda/datasets/drug_affinity_dataset.py [hunk 2]
```diff
@@ -19,6 +22,7 @@ def __init__(
         drug_affinity_filepath: str,
         smi_filepath: str,
         protein_filepath: str,
+        column_names: Tuple[str] = ['label', 'ligand_name', 'sequence_id'],
```
Comment (C-nit) [inline pytoda/datasets/drug_affinity_dataset.py]: Since we are free to define the order here, I would use the "natural" one as they appear in our files. Usually a label/class column comes last, so it's not only us. This would make the test a little more readable by getting rid of `np.roll`
```suggestion
        column_names: Tuple[str] = ['ligand_name', 'sequence_id', 'label'],
```

File: pytoda/datasets/drug_affinity_dataset.py [hunk 3]
```diff
@@ -52,12 +56,14 @@ def __init__(
...
-                with an index and three header columns named: "ligand_name",
-                "sequence_id", "label".
+            drug_affinity_filepath (str): path to drug affinity .csv file. Currently,
+                the only supported format is .csv, with an index and three header
+                columns named as specified in column_names.
             smi_filepath (str): path to .smi file.
             protein_filepath (str): path to .smi or .fasta file.
+            column_names (Tuple[str]): Names of columns in data files to retrieve
+                labels, ligands and protein name respectively.
+                Defaults to ['label', 'ligand_name', 'sequence_id'].
```
Comment (C-nit) [inline pytoda/datasets/drug_affinity_dataset.py]:
```suggestion
                Defaults to ['ligand_name', 'sequence_id', 'label'].
```

File: pytoda/datasets/drug_affinity_dataset.py [hunk 4]
```diff
@@ -129,6 +135,14 @@ def __init__(
         self.device = device
         # backend
         self.backend = backend
+
+        if not isinstance(column_names, Iterable):
+            raise TypeError(f'Column names was {type(column_names)}, not Iterable.')
+        if not len(column_names) == 3:
+            raise ValueError(f'Please pass 3 column names not {len(column_names)}')
+        self.column_names = column_names
+        self.label_name, self.drug_name, self.protein_name = self.column_names
```
Comment (C-nit) [inline pytoda/datasets/drug_affinity_dataset.py]:
```suggestion
        self.drug_name, self.protein_name, self.label_name = self.column_names
```

File: pytoda/datasets/drug_sensitivity_dataset.py [hunk 5]
```diff
@@ -18,6 +18,7 @@ def __init__(
         drug_sensitivity_filepath: str,
         smi_filepath: str,
         gene_expression_filepath: str,
+        column_names: Tuple[str] = ['IC50', 'drug', 'cell_line'],
```
Comment (C-nit) [inline pytoda/datasets/drug_sensitivity_dataset.py]:
```suggestion
        column_names: Tuple[str] = ['drug', 'cell_line', 'IC50'],
```

File: pytoda/datasets/drug_sensitivity_dataset.py [hunk 6]
```diff
@@ -52,14 +53,16 @@ def __init__(
...
+            drug_sensitivity_filepath (str): path to drug sensitivity .csv file.
+                Currently, the only supported format is .csv, with an index and three
+                header columns named as specified in the column_names argument.
             smi_filepath (str): path to .smi file.
             gene_expression_filepath (str): path to gene expression .csv file.
                 Currently, the only supported format is .csv,
                 with an index and header columns containing gene names.
+            column_names (Tuple[str]): Names of columns in data files to retrieve
+                labels, ligands and protein name respectively.
+                Defaults to ['IC50', 'drug', 'cell_line'].
```
Comment (C-nit) [inline pytoda/datasets/drug_sensitivity_dataset.py]:
```suggestion
                Defaults to ['drug', 'cell_line', 'IC50'].
```

File: pytoda/datasets/drug_sensitivity_dataset.py [hunk 7]
```diff
@@ -129,6 +132,14 @@ def __init__(
         self.device = device
         # backend
         self.backend = backend
+
+        if not isinstance(column_names, Iterable):
+            raise TypeError(f'Column names was {type(column_names)}, not Iterable.')
+        if not len(column_names) == 3:
+            raise ValueError(f'Please pass 3 column names not {len(column_names)}')
+        self.column_names = column_names
+        self.label_name, self.drug_name, self.cell_name = self.column_names
```
Comment (C-nit) [inline pytoda/datasets/drug_sensitivity_dataset.py]:
```suggestion
        self.drug_name, self.cell_name, self.label_name = self.column_names
```

File: pytoda/datasets/tests/test_drug_affinity_dataset.py [hunk 8]
```diff
@@ -34,21 +34,28 @@ class TestDrugAffinityDatasetEagerBackend(unittest.TestCase):
...
-                        smiles_file.filename,
-                        protein_sequence_file.filename,
-                        backend=self.backend,
-                    )
+        for column_content in COLUMN_NAMES:
+            self.drug_affinity_content = os.linesep.join(
+                [column_content, DRUG_AFFINITY_CONTENT]
+            )
+            # Bring column names in right order
+            column_names = list(np.roll(column_content.split(',')[1:], 1))
```
Comment (C-nit) [inline pytoda/datasets/tests/test_drug_affinity_dataset.py]:
```suggestion
```

File: pytoda/datasets/tests/test_drug_affinity_dataset.py [hunk 9]
```diff
@@ -34,21 +34,28 @@ class TestDrugAffinityDatasetEagerBackend(unittest.TestCase):
...
-                with TestFileContent(
-                    self.protein_sequence_content
-                ) as protein_sequence_file:
-                    self.drug_affinity_dataset = DrugAffinityDataset(
-                        drug_affinity_file.filename,
-                        smiles_file.filename,
-                        protein_sequence_file.filename,
-                        backend=self.backend,
-                    )
+        for column_content in COLUMN_NAMES:
```
Comment (C-nit) [inline pytoda/datasets/tests/test_drug_affinity_dataset.py]:
```suggestion
        for column_names in COLUMN_NAMES:
```

File: pytoda/datasets/tests/test_drug_affinity_dataset.py [hunk 10]
```diff
@@ -126,21 +137,28 @@ class TestDrugAffinityDatasetLazyBackend(TestDrugAffinityDatasetEagerBackend):
...
```
Comment (C-nit) [inline pytoda/datasets/tests/test_drug_affinity_dataset.py]:
```suggestion
        for column_names in COLUMN_NAMES:
```

File: pytoda/datasets/tests/test_drug_affinity_dataset.py [hunk 11]
```diff
@@ -126,21 +137,28 @@ class TestDrugAffinityDatasetLazyBackend(TestDrugAffinityDatasetEagerBackend):
...
+            self.drug_affinity_content = os.linesep.join(
+                [column_content, DRUG_AFFINITY_CONTENT]
+            )
+            # Bring column names in right order
+            column_names = list(np.roll(column_content.split(',')[1:], 1))
```
Comment (C-nit) [inline pytoda/datasets/tests/test_drug_affinity_dataset.py]:
```suggestion
```

File: pytoda/datasets/tests/test_drug_sensitivity_dataset.py [hunk 12]
```diff
@@ -34,22 +34,31 @@ class TestDrugSensitivityDatasetEagerBackend(unittest.TestCase):
...
-                    self.gene_expression_content
-                ) as gene_expression_file:
-                    self.drug_sensitivity_dataset = DrugSensitivityDataset(
-                        drug_sensitivity_file.filename,
-                        smiles_file.filename,
-                        gene_expression_file.filename,
-                        gene_expression_kwargs={'pandas_dtype': {'genes': str}},
-                        backend=self.backend,
-                    )
+        for column_content in COLUMN_NAMES:
```
Comment (C-nit) [inline pytoda/datasets/tests/test_drug_sensitivity_dataset.py]:
```suggestion
        for column_names in COLUMN_NAMES:
```

File: pytoda/datasets/tests/test_drug_sensitivity_dataset.py [hunk 13]
```diff
@@ -34,22 +34,31 @@ class TestDrugSensitivityDatasetEagerBackend(unittest.TestCase):
...
+            self.drug_sensitivity_content = os.linesep.join(
+                [column_content, DRUG_SENSITIVITY_CONTENT]
+            )
+            # Bring column names in right order
+            column_names = list(np.roll(column_content.split(',')[1:], 1))
```
Comment (C-nit) [inline pytoda/datasets/tests/test_drug_sensitivity_dataset.py]:
```suggestion
```

File: pytoda/datasets/tests/test_drug_sensitivity_dataset.py [hunk 14]
```diff
@@ -109,22 +122,31 @@ class TestDrugSensitivityDatasetLazyBackend(
...
-                    self.gene_expression_content
-                ) as gene_expression_file:
-                    self.drug_sensitivity_dataset = DrugSensitivityDataset(
-                        drug_sensitivity_file.filename,
-                        smiles_file.filename,
-                        gene_expression_file.filename,
-                        gene_expression_kwargs={'pandas_dtype': {'genes': str}},
-                        backend=self.backend,
-                    )
+        for column_content in COLUMN_NAMES:
```
Comment (C-nit) [inline pytoda/datasets/tests/test_drug_sensitivity_dataset.py]:
```suggestion
        for column_names in COLUMN_NAMES:
```

File: pytoda/datasets/tests/test_drug_sensitivity_dataset.py [hunk 15]
```diff
@@ -109,22 +122,31 @@ class TestDrugSensitivityDatasetLazyBackend(
...
+            self.drug_sensitivity_content = os.linesep.join(
+                [column_content, DRUG_SENSITIVITY_CONTENT]
+            )
+            # Bring column names in right order
+            column_names = list(np.roll(column_content.split(',')[1:], 1))
```
Comment (C-nit) [inline pytoda/datasets/tests/test_drug_sensitivity_dataset.py]:
```suggestion
```

File: pytoda/smiles/tests/test_smiles_language.py [hunk 16]
```diff
@@ -78,6 +80,23 @@ def test_add_token(self) -> None:
...
+                torch.multiprocessing.spawn(
+                    s, args=(), nprocs=1, join=True, daemon=False, start_method='spawn'
+                )
+            except AttributeError:
+                # Catch this exception since it is likely caused by Pickling Error
+                raise AttributeError()
+            except Exception:
+                # Other errors are okay since s is not callable and raises a TypeError
+                # But spawn doesn't allow to catch error types, so we use bare Exception
+                pass
```
Comment (C-nit) [inline pytoda/smiles/tests/test_smiles_language.py:98]: This Error handling does not look like we can trust the test.
I played a bit with it, and I can make the process succeed and also catch its return for assertion, but then I can't reproduce the error:

```python
from multiprocessing import Pipe
import torch


class Foo:
    def __init__(self):
        self.set_lam()

    def set_lam(self):
        self.lam = lambda x: x

    def call_lam(self):
        x = self.lam(42)
        print(x)
        return x


foo = Foo()


def worker(i, sender):
    sender.send(foo.call_lam())

if __name__ == '__main__':
    receiver, sender = Pipe(False)
    torch.multiprocessing.spawn(worker, args=(sender,))
    print(f'received {receiver.recv()}')
```
the above prints 
```console
42
received 42
```
Do you have an idea? We need to be able to make the test fail, else we can't claim to reproduce the issue.
I'll patch something similar here to see how it behaves with smiles language.

```suggestion
        from multiprocessing import Pipe
        for start_stop in [False, True]:
            s = SMILESTokenizer(add_start_and_stop=start_stop)

            def worker(i, smiles, sender):
                sender.send(s.smiles_to_token_indexes(smiles))

            receiver, sender = Pipe(False)
            torch.multiprocessing.spawn(
                s, args=('CCO', sender), nprocs=1, join=True, daemon=False, start_method='spawn'
            )
            token_indexes = receiver.recv()
            self.assertEqual(start_stop, len(token_indexes) == 3)
```

File: requirements.txt [hunk 17]
```diff
@@ -9,3 +9,4 @@ upfp==0.0.5
 SmilesPE>=0.0.3
 pyfaidx==0.5.9.1
 pubchempy==1.0.4
+coverage==5.3
```
Comment (C-nit) [inline requirements.txt]: `coverage` is not a requirement.
We could start a `dev_requirements.txt` or so, and also add black and flake8 to it, and also add it into the conda.yml.

**Review by C-nit — CHANGES_REQUESTED** [review]
Nice work!
We need to look at the spawn thingy before merging, though.
The other suggested changes you can decide to ignore if you feel strongly about the order.

File: requirements.txt (hunk 17 above)
Comment (jannisborn) [reply requirements.txt]: I moved it away from here to the `Dockerfile`.

**Review by C-nit — APPROVED** [review]
